src/main/kotlin/org/rust/lang/core/types/infer/TypeInference.kt

```diff
@@ -0,0 +1,434 @@
+ */
+
+package org.rust.lang.core.types.infer
+
+import com.intellij.psi.PsiElement
+import org.rust.ide.utils.isNullOrEmpty
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.*
+import org.rust.lang.core.resolve.*
+import org.rust.lang.core.resolve.ref.resolveFieldExprReferenceWithReceiverType
+import org.rust.lang.core.resolve.ref.resolveMethodCallReferenceWithReceiverType
+import org.rust.lang.core.types.ty.*
+import org.rust.lang.core.types.type
+import org.rust.utils.toIntOrNull
+
+fun inferTypesIn(fn: RsFunction): RsInferenceContext {
+    val ctx = RsInferenceContext()
+    TypeInferenceRecursionGuard.guarded { ctx.infer(fn) }
+    return ctx
+}
+
+class RsInferenceContext {
+    val bindings: MutableMap<RsPatBinding, Ty> = mutableMapOf()
+    val exprTypes: MutableMap<RsExpr, Ty> = mutableMapOf()
+
+    fun infer(fn: RsFunction) {
+        extractParameterBindings(fn)
+
+        val block = fn.block
+        if (block != null) {
+            RsFnInferenceContext(this).walkBlock(block)
+        }
+    }
+
+    private fun extractParameterBindings(fn: RsFunction) {
+        for(param in fn.valueParameters) {
+            collectBindings(param.pat, param.typeReference?.type ?: TyUnknown)
+        }
+    }
+
+    fun getExprType(expr: RsExpr): Ty {
+        return exprTypes[expr] ?: TyUnknown
+    }
+
+    fun getBindingType(binding: RsPatBinding): Ty {
+        return bindings[binding] ?: TyUnknown
+    }
+
+    fun writeTy(psi: RsExpr, ty: Ty) {
+        exprTypes[psi] = ty
+    }
+
+    fun collectBindings(pattern: RsPat?, baseType: Ty) {
+        bindings += collectBindingsNullable(pattern, baseType);
+    }
+
+    override fun toString(): String {
+        return "RsInferenceContext(bindings=$bindings, exprTypes=$exprTypes)"
+    }
+}
+
+private class RsFnInferenceContext(private val ctx: RsInferenceContext) {
+    private val RsExpr.ty: Ty get() = ctx.getExprType(this)
+    private val RsBlock.ty: Ty get() = expr?.ty ?: TyUnit
+    private val RsStructLiteralField.type: Ty get() = resolveToDeclaration?.typeReference?.type ?: TyUnknown
+
+    fun walkBlock(block: RsBlock) {
+        for (stmt in block.stmtList) walk(stmt)
+        walkNullable(block.expr)
+    }
+
+    private fun walkNullable(psi: PsiElement?) {
+        if(psi != null) walk(psi)
+    }
+
+    private fun walkChildren(psi: PsiElement) {
+        for (psi1 in psi.children) walk(psi1)
+    }
+
+    private fun walk(psi: PsiElement) {
+        when (psi) {
+            is RsLetDecl -> {
+                walkNullable(psi.expr)
+                ctx.collectBindings(psi.pat, psi.typeReference?.type ?: psi.expr?.ty ?: TyUnknown)
+            }
+            is RsCondition -> {
+                walkNullable(psi.expr)
+                ctx.collectBindings(psi.pat, psi.expr.ty)
+            }
+            is RsMatchArm -> {
+                for (pat in psi.patList)
+                    ctx.collectBindings(pat, psi.parentOfType<RsMatchExpr>()?.expr?.ty ?: TyUnknown)
```

Comment:
Can we have braces here and in similar cases elsewhere ? :) 

https://www.imperialviolet.org/2014/02/22/applebug.html